lib/streamlit/components/v1/base_component_registry.py

```diff
@@ -0,0 +1,71 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Protocol
+
+from streamlit.components.v1.custom_component import CustomComponent
+
+
+class BaseComponentRegistry(Protocol):
+    """Interface for implementing an alternative ComponentRegistry."""
+
+    @abstractmethod
+    def register_component(self, component: CustomComponent) -> None:
+        """Register a CustomComponent.
+
+        Parameters
+        ----------
+        component : CustomComponent
+            The component to register.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_component_path(self, name: str) -> str | None:
+        """Return the filesystem path for the component with the given name.
+
+        If no such component is registered, or if the component exists but is
+        being served from a URL, return None instead.
```

Comment:
I agree, though I am hesitant to make changes to the API that has existed so far since I am not sure whether someone relies on this being `None` in the case where the component is served via a URL. Hence, I would suggest that we look into modifying the API in the context of the broader work on CustomComponents and try to keep this PR more on the re-factoring side. Does this make sense to you?